### PR TITLE
Remove size one spec axis for MaxText sharding

### PR DESCRIPTION
**Pull request by NuojCheng**
### Description

This PR addresses issues related to redundant partition axes of size one in sharding specifications. These redundant axes can be misleading, complicate debugging, and have caused confusion with `jax.lax.pvary`. They also prevent the use of `check_vma=True` in `jax.shard_map`.

For example, if we only use FSDP (`ici_fsdp_parallelism=-1`), the sharding specification for an activation tensor would be:

**Before this PR:**
```python
> flax.linen.logical_to_mesh_axes(("activation_batch", "activation_length", "activation_embed"), rules)
> P(('data', 'fsdp', 'fsdp_transpose', 'expert'), ('sequence', 'context', 'expert'), ('tensor', 'tensor_transpose'))
```

**After this PR:**
```python
> MaxText.sharding.logical_to_mesh_axes(("activation_batch", "activation_length", "activation_embed"), mesh, rules)
> P("fsdp", None, None)
```

To solve this, this PR introduces two new utility functions in `MaxText.sharding`:
1. `logical_to_mesh_axes`: This function is a wrapper around `linen.logical_to_mesh_axes` that also removes mesh axes of size one from the resulting PartitionSpec. This simplifies the sharding specification and avoids potential issues with JAX APIs.
2. `create_sharding`: This function simplifies the creation of `NamedSharding` objects by combining the mesh and logical axis names into a single call.

The PR then refactors the existing codebase to use these new utility functions, replacing direct calls to `linen.logical_to_mesh_axes` and manual creation of `NamedSharding` objects.

# Tests

This PR includes targeted testing to validate the fix for the TP issue with MoE models in b/460216386 and b/458796896. The following tests confirm that the sharding strategy now works correctly with `ici_tensor_parallelism=2`.

| Model | Test Command | Log |
| :--- | :--- | :--- |
| **Deepseek3-test** | `smoke_train model_name=deepseek3-test per_device_batch_size=1 shard_mode=auto ici_tensor_parallelism=2 use_tokamax_gmm=true` | [View Log](https://paste.googleplex.com/5716110378926080) |
| **GPT-OSS-20b** (single layer) | `smoke_train model_name=gpt-oss-20b per_device_batch_size=1 shard_mode=auto ici_tensor_parallelism=2 use_tokamax_gmm=true` | [View Log](https://paste.googleplex.com/5275592662581248) |

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/sharding.py [hunk 1]
```diff
@@ -30,7 +30,8 @@
 
 def get_input_data_sharding(config, mesh):
   """Get the input data sharding for the model"""
```
[the remark this answers is not in the document]
Comment (NuojCheng) [reply src/MaxText/sharding.py:32]: Noooo it won't work. `create_sharding` only works under the context of logical sharding rules, which we dont have for the `get_input_data_sharding` function. I will ignore this suggestion.

File: src/MaxText/sharding.py [hunk 2]
```diff
@@ -48,10 +49,40 @@ def maybe_shard_with_logical(inputs, logical_axes, mesh, shard_mode):
...
   """
-  named_sharding = NamedSharding(mesh, nn.logical_to_mesh_axes(logical_axes))
+  named_sharding = create_sharding(mesh, logical_axes)
   return maybe_shard_with_name(inputs, named_sharding, shard_mode)
 
 
+def remove_size_one_mesh_axis(spec, mesh) -> P:
+  """
+  Removes mesh axes from a PartitionSpec (P) where the axis size is 1.
+
```
[the remark this answers is not in the document]
Comment (NuojCheng) [reply src/MaxText/sharding.py:62]: love it but I copied-paste from jax, maybe I should just keep whatever it is?

**Review by richjames0 — APPROVED** [review]
lgtm

**Comment by NuojCheng** [comment]
> Nice fix!!! Did you see any perf changes before/after this change for those test models when `use_tokamax_gmm=False`?

They are matched for GPT-OSS: https://diff.googleplex.com/#key=EGlo0V8OfzVa

**Review by RissyRan — APPROVED** [review]

File: src/MaxText/sharding.py [hunk 3]
```diff
@@ -46,16 +46,48 @@ def maybe_shard_with_name(inputs, named_sharding, shard_mode):
...
+  This is a common optimization to simplify sharding by excluding redundant axes.
+  Function originally from jax._src.core:
+  https://github.com/jax-ml/jax/blob/main/jax/_src/core.py
```
Comment (RissyRan) [inline src/MaxText/sharding.py:65]: Nit: use this should work: https://github.com/jax-ml/jax/blob/f37438779256895cb242f878d15c444ae8d16243/jax/_src/core.py#L2079-L2088

https://screenshot.googleplex.com/9nSTXN7o9rrTqp6

Comment (NuojCheng) [reply src/MaxText/sharding.py:65]: I tried this earlier in https://github.com/AI-Hypercomputer/maxtext/pull/2579#discussion_r2587791250 but somehow the link expires over time, probably because the commit attached with the link gets revoked? 

Another way to refer is probably do https://github.com/jax-ml/jax/blob/main/jax/_src/core.py#L2079-L2088, but line numbers will change over time. Probably I should just keep what as it is, just refer to the file instead of the lines.
